Adding Node Health Management wiki page (#21)
ocp-power-automation/extra: docs/Node_Health_Management.md

# Managing Node Health

A node provides the runtime environments for containers. Each node in a Kubernetes cluster has the required services to be managed by the master. Nodes also have the required services to run pods, including the container runtime, a kubelet, and a service proxy.

OpenShift Container Platform creates nodes from a cloud provider, physical systems, or virtual systems. Kubernetes interacts with node objects that are a representation of those nodes. The master uses the information from node objects to validate nodes with health checks. A node is ignored until it passes the health checks, and the master continues checking nodes until they are valid.
We can manage nodes in our instance using the CLI. When we perform node management operations, the CLI interacts with node objects that are representations of actual node hosts. The master uses the information from node objects to validate nodes with health checks.

You can review cluster node health status, resource consumption statistics, and node logs using the below commands. Additionally, you can query kubelet status on individual nodes.

## 1. Node reporting NotReady
***

> The mentioned problem was seen on 4.5.4 running on KVM/libvirt, but this could happen and be applied on other platforms.Some nodes in this 3-master/5-worker cluster start showing "NotReady" state with "Kubelet stopped posting node status" in `oc describe node` output.

Reviewer: ```suggestion
The mentioned problem was seen on 4.5.4 running on KVM/libvirt, but this could happen and be applied on other platforms.
Some nodes in this 3-master/5-worker cluster start showing "NotReady" state with "Kubelet stopped posting node status" in `oc describe node` output.
```

Space was missing between the sentences, and I like giving each sentence a line in .md format (but will be rendered naturally by `mkdocs`) for easier maintenance later.

Feb 25 11:33:12 master01 etcd[1103]: got unexpected response error (etcdserver: request timed out)
Feb 25 11:33:12 master01 etcd[1103]: got unexpected response error (etcdserver: request timed out)
[root@master00 ~]# systemctl status etcd
● etcd.service - Etcd Server
   Loaded: loaded (/usr/lib/systemd/system/etcd.service; enabled; vendor preset: disabled)
   Active: active (running) since Thu 2016-02-25 08:43:55 CST; 4h 32min ago
 Main PID: 1097 (etcd)
   CGroup: /system.slice/etcd.service
           └─1097 /usr/bin/etcd --name=master00.ose.example.com --data-dir=/var/lib/etcd/ --lis...

Feb 25 11:40:25 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
Feb 25 11:40:55 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
Feb 25 11:41:25 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
Feb 25 11:41:55 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
Feb 25 11:42:25 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
```

**Solution**

In most cases restarting etcd one at a time on each etcd host resolves the issue

> # `systemctl restart etcd`

Reviewer: ```suggestion
$ `sudo systemctl restart etcd`
```

The root prompt (#) was interpreted as `<h1>`, and this is the easy way out.

> #  **Managing Node Health**

Reviewer: ```suggestion
# Managing Node Health
```
Let's just have the `<h1>` does its thing, no extra formatting.

A node provides the runtime environments for containers. Each node in a Kubernetes cluster has the required services to be managed by the master. Nodes also have the required services to run pods, including the container runtime, a kubelet, and a service proxy.

OpenShift Container Platform creates nodes from a cloud provider, physical systems, or virtual systems. Kubernetes interacts with node objects that are a representation of those nodes. The master uses the information from node objects to validate nodes with health checks. A node is ignored until it passes the health checks, and the master continues checking nodes until they are valid.
We can manage nodes in our instance using the CLI. When we perform node management operations, the CLI interacts with node objects that are representations of actual node hosts. The master uses the information from node objects to validate nodes with health checks.

You can review cluster node health status, resource consumption statistics, and node logs using the below commands. Additionally, you can query kubelet status on individual nodes.

> **1. Node reporting Not Ready**

Reviewer: ```suggestion
## 1. Node reporting NotReady
```

By using those headings, they start showing up on the left navigation (after `mkdocs` does its magic).

***

> In the event that you run
> ```oc get nodes``` and the result shows that the node is not ready

Reviewer: ```suggestion
In the event that you run `oc get nodes` and the result shows that the node is not ready
```

Start by investigating the atomic-openshift-node service:

> ```

Reviewer: ```suggestion
```text
```

This would prevent coloring in the code block.

Oct 29 07:56:00 master0.ocp-support-suad.cp.fyre.ibm.com atomic-openshift-node[2412]: I0225 08:37:31.956434    2412 proxier.go:352] Setting endpoints for "default/kubernetes:https" to [192.168.200.50:8443 192.168.200.51:8443 192.168.200.52:8443]
Oct 29 07:56:00 master0.ocp-support-suad.cp.fyre.ibm.com atomic-openshift-node[2412]: I0225 08:37:31.956476    2412 proxier.go:429] Not syncing iptables until Services and Endpoints have been received from master
Oct 29 07:56:00 master0.ocp-support-suad.cp.fyre.ibm.com atomic-openshift-node[2412]: I0225 08:37:31.965155    2412 proxier.go:294] Adding new service "default/kubernetes:https" at 172.50.0.1:443/TCP
Oct 29 07:56:00 master0.ocp-support-suad.cp.fyre.ibm.com atomic-openshift-node[2412]: I0225 08:37:31.965358    2412 proxier.go:294] Adding new service "default/kubernetes:dns" at 172.50.0.1:53/UDP
Oct 29 07:56:00 master0.ocp-support-suad.cp.fyre.ibm.com atomic-openshift-node[2412]: I0225 08:37:31.965450    2412 proxier.go:294] Adding new service "default/kubernetes:dns-tcp" at 172.50.0.1:53/TCP
```
This will resolve the problem most of the times and ```oc get nodes``` will give the positive result.

> **/var/log/messages** can sometimes shed some additional light if the problem is not resolved by restarting the atomic-openshift-node service.

Reviewer: ```suggestion
`/var/log/messages` can sometimes shed some additional light if the problem is not resolved by restarting the atomic-openshift-node service.
```

Cosmetic change, but we have a bigger problem here, my bad, and this `atomic-openshift-node` service is an OKD implementation, I believe. We don't have that in OCP. Either we try to rewrite this section with OCP equivalent, or we just publish this version and quickly try to replace after.

  **2. Querying the kubelet’s status on a node**
***

The kubelet is managed using a systemd service on each node. Review the kubelet’s status by querying the kubelet systemd service within a debug Pod.

> a. Start a debug Pod for a node: `$ oc debug node/node_name`

Reviewer: ```suggestion
a. Start a debug Pod for a node: $ `oc debug node/node_name`
```

> b.  Set /host as the root directory within the debug shell. The debug Pod mounts the host’s root file system in /host within the Pod. By changing the root directory to /host, you can run binaries contained in the host’s executable paths:`# chroot /host`

Reviewer: ```suggestion
b.  Set /host as the root directory within the debug shell. The debug Pod mounts the host’s root file system in /host within the Pod. By changing the root directory to /host, you can run binaries contained in the host’s executable paths: # `chroot /host`
```

> c.  Check whether the kubelet systemd service is active on the node:
>  ```# systemctl is-active kubelet```

Reviewer: ```suggestion
c.  Check whether the kubelet systemd service is active on the node: # `systemctl is-active kubelet`
```

> d. Output a more detailed kubelet.service status summary:
> ``` # systemctl status kubelet```

Reviewer: ```suggestion
d. Output a more detailed kubelet.service status summary: # `systemctl status kubelet`
```

**_Status on CLI_**

> ```

Reviewer: ```suggestion
```text
```

Aug 28 11:35:41 master0.ocp-support-suad.cp.fyre.ibm.com hyperkube[1349]: I0828 11:35:41.828681 1349 atomic_writer.go:157] pod openshift-kube-scheduler-operator/openshift-kube-scheduler-operator-769d7b>
Aug 28 11:35:41 master0.ocp-support-suad.cp.fyre.ibm.com hyperkube[1349]: I0828 11:35:41.828717 1349 operation_generator.go:657] MountVolume.SetUp succeeded for volume "serving-cert" (UniqueName: "kube>
Aug 28 11:35:41 master0.ocp-support-suad.cp.fyre.ibm.com hyperkube[1349]: I0828 11:35:41.837829 1349 atomic_writer.go:157] pod openshift-image-registry/node-ca-vhm68 volume serviceca: no update require>
Aug 28 11:35:41 master0.ocp-support-suad.cp.fyre.ibm.com hyperkube[1349]: I0828 11:35:41.837926 1349 operation_generator.go:657] MountVolume.SetUp succeeded for volume "serviceca" (UniqueName: "kuberne>
Aug 28 11:35:41 master0.ocp-support-suad.cp.fyre.ibm.com hyperkube[1349]: I0828 11:35:41.970203 1349 exec.go:60] Exec probe response: ""
Aug 28 11:35:41 master0.ocp-support-suad.cp.fyre.ibm.com hyperkube[1349]: I0828 11:35:41.970272 1349 prober.go:133] Liveness probe for "ovs-p2qkz_openshift-sdn(6d62e5db-053b-4573-95ce-d7d478f578d9):ope>

```

>  **3. Node-logs: Get logs for NetworkManager**

Reviewer: ```suggestion
 ## 3. Node-logs: Get logs for NetworkManager
```

***

The below command will give the logs of the network manager of a particular node of a cluster.

>                 
> ```  oc adm node-logs --role master -u NetworkManager.service```
> 
> 

Reviewer: ```suggestion
$ `oc adm node-logs --role master -u NetworkManager.service`
```

> ```

Reviewer: ```suggestion
```text
```

Sep 06 04:01:37.044593 master2.ocp-support-suad.cp.fyre.ibm.com NetworkManager[1163]: <info>  [1599364897.0445] dhcp4 (enp0s3): option requested_static_routes => '1'
Sep 06 04:01:37.044613 master2.ocp-support-suad.cp.fyre.ibm.com NetworkManager[1163]: <info>  [1599364897.0445] dhcp4 (enp0s3): option requested_subnet_mask => '1'
Sep 06 04:01:37.044637 master2.ocp-support-suad.cp.fyre.ibm.com NetworkManager[1163]: <info>  [1599364897.0446] dhcp4 (enp0s3): option requested_time_offset => '1'
Sep 06 04:01:37.044658 master2.ocp-support-suad.cp.fyre.ibm.com NetworkManager[1163]: <info>  [1599364897.0446] dhcp4 (enp0s3): option requested_wpad       => '1'
Sep 06 04:01:37.044734 master2.ocp-support-suad.cp.fyre.ibm.com NetworkManager[1163]: <info>  [1599364897.0446] dhcp4 (enp0s3): option routers              => '10.17.64.1'
Sep 06 04:01:37.044759 master2.ocp-support-suad.cp.fyre.ibm.com NetworkManager[1163]: <info>  [1599364897.0447] dhcp4 (enp0s3): option subnet_mask          => '255.255.224.0'
Sep 06 04:01:37.044779 master2.ocp-support-suad.cp.fyre.ibm.com NetworkManager[1163]: <info>  [1599364897.0447] dhcp4 (enp0s3): state changed extended -> extended
```

> **4. Nodes report ready but ETCD health check fails**

Reviewer: ```suggestion
## 4. Nodes report ready but ETCD health check fails
```

***

> ```

Reviewer: ```suggestion
```text
```

[root@master02 ~]#  etcdctl -C https://master00.ose.example.com:2379,https://master01.ose.example.com:2379,https://master01.ose.example.com:2379 --ca-file=/etc/origin/master/master.etcd-ca.crt     --cert-file=/etc/origin/master/master.etcd-client.crt     --key-file=/etc/origin/master/master.etcd-client.key cluster-health
member e0e2c123213680f is healthy: got healthy result from https://192.168.200.50:2379
member 64f1077d838e039c is healthy: got healthy result from https://192.168.200.51:2379
member a9e031ea9ce2a521 is unhealthy: got unhealthy result from https://192.168.200.52:2379
```

In the event that the health check fails check the status of etcd you could see one or a combination of the following:

> ```

Reviewer: ```suggestion
```text
```

Feb 25 11:33:12 master01 etcd[1103]: got unexpected response error (etcdserver: request timed out)
Feb 25 11:33:12 master01 etcd[1103]: got unexpected response error (etcdserver: request timed out)
[root@master00 ~]# systemctl status etcd
● etcd.service - Etcd Server
   Loaded: loaded (/usr/lib/systemd/system/etcd.service; enabled; vendor preset: disabled)
   Active: active (running) since Thu 2016-02-25 08:43:55 CST; 4h 32min ago
 Main PID: 1097 (etcd)
   CGroup: /system.slice/etcd.service
           └─1097 /usr/bin/etcd --name=master00.ose.example.com --data-dir=/var/lib/etcd/ --lis...

Feb 25 11:40:25 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
Feb 25 11:40:55 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
Feb 25 11:41:25 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
Feb 25 11:41:55 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
Feb 25 11:42:25 master00 etcd[1097]: the connection to peer a9e031ea9ce2a521 is unhealthy
```

**Solution**

In most cases restarting etcd one at a time on each etcd host resolves the issue

> `systemctl restart etcd`

Reviewer: ```suggestion
 # `systemctl restart etcd`
```

> Reference Link - http://v1.uncontained.io/playbooks/troubleshooting/troubleshooting_guide.html

Reviewer: ```suggestion
Reference Link - <http://v1.uncontained.io/playbooks/troubleshooting/troubleshooting_guide.html>
```

> **5. File System tracking(a complete monitoring for disk usage)**

Reviewer: ```suggestion
## 5. File System tracking(a complete monitoring for disk usage)
```

***

To monitor filesystems and more precisely the overall space remaining on various filesystems , 
the node exporter exports two metrics to retrieve such statistics :

> a. `node_filesystem_avail_bytes `

Reviewer: ```suggestion
a. `node_filesystem_avail_bytes `  
```

adding 2 spaces at the end of line creates a line-break.

b. `node_filesystem_size_bytes`

> And this will give us the overall filesystem usage by device or by mountpoint :

Reviewer: ```suggestion
Use below query in Prometheus for the overall filesystem usage by device or by mountpoint :
```

`( 1 - node_filesystem_avail_bytes /node_filesystem_size_bytes )*100`

>  **6. Read & Write Latencies**

Reviewer: ```suggestion
## 6. Read & Write Latencies
```

***

Another great metric for monitoring is the read and write latencies on our disks.
The node exporter exports multiple metrics to compute it.
On the read side, we have:

> a.`node_disk_read_time_seconds_total`

Reviewer: ```suggestion
a.`node_disk_read_time_seconds_total`  
```

b.`node_disk_reads_completed_total`

On the write side, we have:

> a.`node_disk_write_time_seconds_total`

Reviewer: ```suggestion
a.`node_disk_write_time_seconds_total`  
```

b.`node_disk_writes_completed_total`

> If we compute rates for the two metrics, and divide one by the other, we are able to compute the latency or the time that your disk takes in order to complete such operations.

Reviewer: ```suggestion
If we compute rates for the two metrics, and divide one by the other, we are able to compute the latency or the time that your disk takes in order to complete such operations.
Use below query in Prometheus for the disk latency :
```
